### PR TITLE
GC-55366 custom cache function support

### PR DESCRIPTION
**Pull request by mwhearty**
Add support for custom cache function for rules.

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -29,8 +29,9 @@ export declare class ILogicRule {
   resolve(parent, args, ctx, info, options: IOptions): Promise<IRuleResult>
 }
 
+export type ICacheFunction = (parent: any, args: any) => string
```
Comment (hmak3d) [inline src/types.ts:32]: Should `args` (2nd parameter) not be `any` type but instead type matching object-hash `hash()` [option](https://www.npmjs.com/package/object-hash#hashvalue-options)?

File: src/rules.ts [hunk 2]
```diff
@@ -161,6 +161,9 @@ export class Rule implements IRule {
       case 'no_cache': {
         return `${this.name}-${Math.random()}`
       }
+      default: {
+        return `${this.name}-${this.cache(parent, args)}`
```
Comment (hmak3d) [inline src/rules.ts:165]: Can `this.cache` be null or non-function?

File: package.json [hunk 3]
```diff
@@ -1,7 +1,7 @@
 {
-  "name": "graphql-shield",
+  "name": "grabcad-graphql-shield",
```
Comment (hmak3d) [inline package.json:2]: We should rename the repository `grabcad-graphql-shield`

**Review by hmak3d — APPROVED** [review]
Minor tweaks requested but if it's too much work, just :shipit:

File: src/types.ts (hunk 1 above)
Comment (mwhearty) [reply src/types.ts:32]: just matching what this module is doing for typings with the resolver, which has `any` as args.  i agree this module doesnt do typing great

File: src/rules.ts (hunk 2 above)
Comment (mwhearty) [reply src/rules.ts:165]: it cannot, if cache option is falsy it gets set to contextual

File: package.json (hunk 3 above)
Comment (mwhearty) [reply package.json:2]: i think leaving it like it is makes it more obvious that its a fork and not something private.  anyway i expect we will drop this fork when new version of shield comes out with better cache support, there are already some open issues with PRs around this same thing.

**Pull request merged**